[doc][core][compiled graphs] Add doc for profiling (#49879)
ray-project/ray: doc/source/ray-core/compiled-graph/profiling.rst

Profiling
=========

> Ray Compiled Graph provides profiling functionalities to better understand performance

Reviewer: ```suggestion
Ray Compiled Graph provides profiling functionalities to better understand the performance
```

of individual tasks, systems overhead, and performance bottlenecks.

Nsight system profiler
----------------------

> Compiled Graphs build on top of Ray's profiling capabilities, and leverage Nsight

Reviewer 2: ```suggestion
Compiled graphs build on top of Ray's profiling capabilities, and leverage Nsight
```

Author: I think we want to use capital G to be consistent?

Reviewer 2: Yes - I forgot to remove this comment. Good catch!

> Compiled Graphs build on top of Ray's profiling capabilities, and leverage Nsight

Reviewer 2: ```suggestion
Compiled Graph build on top of Ray's profiling capabilities, and leverage Nsight
```

system profiling. 

> To run Nsight Profiling on Compiled Graphs, specify the runtime_env for the involved actors

Reviewer 2: ```suggestion
To run Nsight Profiling on Compiled Graph, specify the runtime_env for the involved actors
```

as described in `Run Nsight on Ray <https://docs.ray.io/en/latest/ray-observability/user-guides/profiling.html#run-nsight-on-ray>`__.

After execution, Compiled Graph generates the profiling results under the `/tmp/ray/session_*/logs/{profiler_name}`
directory.

For fine-grained performance analysis of method calls and system overhead, set the environment variable

> ``RAY_CGRAPH_ENABLE_NVTX_PROFILING=1`` when running the script. For example, for a compiled graph script

Reviewer 2: ```suggestion
``RAY_CGRAPH_ENABLE_NVTX_PROFILING=1`` when running the script. For example, for a Compiled Graph script
```

in ``example.py``, run the following command:

.. testcode::

    RAY_CGRAPH_ENABLE_NVTX_PROFILING=1 python3 example.py

> This leverages the `NVTX library <https://nvtx.readthedocs.io/en/latest/index.html#>`_ to automatically

Reviewer 2: ```suggestion
This command leverages the `NVTX library <https://nvtx.readthedocs.io/en/latest/index.html#>`_ to automatically
```

> annotate all methods called in the execution loops of compiled graph.

Reviewer 2: ```suggestion
annotate all methods that Ray calls in the execution loops of Compiled Graph.
```

Author: So the methods are called in a chain. The outer most caller is Ray, but the inner callers may be customer code.

Reviewer 2: Ah - I didn't realize that. Thanks for the explanation. 

`annotate all methods that Ray or the application code calls in the execution loops of Compiled Graph.`

Reviewer 2: If you don't like my suggestion, just go with your original wording. This is not a blocker.

Author: I think yours align with the doc style better. But I will keep as is, I feel it is a bit more accurate.

> To visualize the profiling results, follow the same instructions as described in 

Reviewer 2: ```suggestion
To visualize the profiling results, follow the instructions in 
```

> `Nsight Profiling Result <https://docs.ray.io/en/latest/ray-observability/user-guides/profiling.html#profiling-result>`__.

Reviewer 2: ```suggestion
`Nsight Profiling Result <../../ray-observability/user-guides/profiling.html#profiling-result>`__.
```

Reviewer 2: Please verify that the relative path is correct. I just took a guess and didn't verify. We prefer to use relative paths so that the links work for future versions.

Author: Thanks. The path is ../ray-observability/user-guides/profiling.md
But we want to specify the specific section, and
../ray-observability/user-guides/profiling.md#profiling-result does not work.

Reviewer 2: Ah, I see. Can you add an anchor in the destination file and use that instead? This is the syntax:
`.. _profiling-result:`

Add that line right above the section title.

Author: Thanks for the suggestion!

I used the syntax of the following to be consistent with existing usage.

```
(profiling-result)=
```

```
:ref:`Nsight Profiling Result <profiling-result>`
```
